### PR TITLE
Add a timeout when acquiring mutex lock

### LINKED ISSUE #4393: mutex lock timeout
In the `lockFactory`, when acquiring a lock, it can get stuck talking to the database. Future calls will block on the mutex.Lock. This can lead to an avalanche of goroutines from the `buildtracker`, `scheduler`, and `lidar` as each tick piles on more work.

We want to add a timeout to the mutex.Lock so that new goroutines will die quicker when they can't fetch the lock.

### PR DESCRIPTION
**Pull request by clarafu**
fixes #4393 

In the lockFactory, when acquiring a lock, it can get stuck talking to the database. Future calls will block on the mutex.Lock. This can lead to an avalanche of goroutines from the buildtracker, scheduler, and lidar as each tick piles on more work.

We want to add a timeout to the mutex.Lock so that new goroutines will die quicker when they can't fetch the lock.

### CODE REVIEW

File: atc/db/lock/lock.go [hunk 1]
```diff
@@ -147,42 +148,44 @@ type LockDB interface {
...
+	mutex  *mutex
 
 	acquired LogFunc
 	released LogFunc
 }
 
 func (l *lock) Acquire() (bool, error) {
-	l.acquireMutex.Lock()
-	defer l.acquireMutex.Unlock()
+	if l.mutex.Lock(time.Second) {
```
Comment (cirocosta) [inline atc/db/lock/lock.go]: Hey, little nit - wdyt of having the `time.Second` here as a constant, for the purpose of being explicit that that is the timeout we're using? 

thanks!

File: atc/db/lock/lock_test.go [hunk 2]
```diff
@@ -104,6 +104,36 @@ var _ = Describe("Locks", func() {
...
+
+			done := make(chan struct{})
+			readyToAcquire := make(chan struct{})
+
+			fakeLockDB := new(lockfakes.FakeLockDB)
+			factory := lock.NewTestLockFactory(fakeLockDB)
+
+			fakeLockDB.AcquireStub = func(id lock.LockID) (bool, error) {
+				select {
+				case <-time.After(time.Minute):
```
Comment (xtremerui) [inline atc/db/lock/lock_test.go]: If default timeout is 1 second, should we make this test to a bit closer to that so if someone extend the default timeout this test could protect that? How about `time.Minute * 2`

File: atc/db/lock/lock.go (hunk 1 above)
Comment (pivotal-jwinters) [reply atc/db/lock/lock.go]: Fixed

File: atc/db/lock/lock_test.go (hunk 2 above)
Comment (pivotal-jwinters) [reply atc/db/lock/lock_test.go]: Updated

**Comment by pivotal-jwinters** [comment]
@cirocosta @xtremerui do either of you have any general thoughts on this change? Any reservations about why doing this might be a bad idea?

**Pull request merged**